### PR TITLE
Log the Courant number

### PR DESCRIPTION
**Pull request by tommbendall**
This introduces a system to log the Courant number, which will be done by default.

Main points:
- a `log_courant` option is passed to the output configuration
- the `CourantNumber` diagnostic is augmented so that it can be evaluated on an expression, and not just the field named "u" (this is used when the transporting velocity isn't exactly the wind field)
- some plumbing has been added to the IO, so that the Courant number diagnostics are set up if needed. The diagnostics are used to evaluated the maximum value (which is logged)
- the calls to log the Courant number are added to the time loop. The `SemiImplicitQuasiNewton` time stepper logs the Courant number every outer loop

### CODE REVIEW

**Comment by tommbendall** [comment]
> The DG0 Courant number based on DG fluxes is a bit more useful for pinpointing exactly which cell things went bad in.

Oh interesting, thanks Colin! How would we go about calculating that?

**Comment by dham** [comment]
> > The DG0 Courant number based on DG fluxes is a bit more useful for pinpointing exactly which cell things went bad in.
> 
> Oh interesting, thanks Colin! How would we go about calculating that?

Looks like your code is already set up to do it. You just set the output space to a DG0 `FunctionSpace`.

File: gusto/io.py [hunk 1]
```diff
@@ -238,6 +238,60 @@ def log_parameters(self, equation):
...
+        """
+        Sets up Courant number diagnostics to be logged.
+
+        Args:
+            state_fields (:class:`StateFields`): the model's field container.
+            name (str, optional): the name of the field to log the Courant
+                number of. Defaults to 'u'.
+            expression (:class:`ufl.Expr`, optional): expression of velocity
+                field to take Courant number of. Defaults to None, in which case
+
```
Comment (jshipton) [inline gusto/io.py:251]: unfinished comment...

File: gusto/io.py [hunk 2]
```diff
@@ -238,6 +238,60 @@ def log_parameters(self, equation):
...
+    def log_courant(self, state_fields, name='u', message=None):
+        """
+        Logs the maximum Courant number value.
+
+        Args:
+            state_fields (:class:`StateFields`): the model's field container.
+            name (str, optional): the name of the field to log the Courant
+                number of. Defaults to 'u'.
+            message (str, optional): an extra message to be logged. Defaults to
```
Comment (jshipton) [inline gusto/io.py:277]: I like the option to add a message!

**Review by jshipton — CHANGES_REQUESTED** [review]
looks good, and very useful - thanks! Just needs that comment finishing off...

**Review by jshipton — APPROVED** [review]

**Comment by colinjcotter** [comment]
On 24 Aug 2023, at 15:26, David A. Ham ***@***.***> wrote:


Looks like your code is already set up to do it. You just set the output space to a DG0 FunctionSpace.

Message ID: ***@***.***>

No, it’s different.
See: https://github.com/firedrakeproject/asQ/blob/master/utils/diagnostics.py<https://github.com/firedrakeproject/asQ/blob/master/utils/diagnostics.py>

**Pull request merged**